toolchain/check/handle_where.cpp

```diff
@@ -17,44 +20,75 @@ auto HandleParseNode(Context& context, Parse::WhereOperandId node_id) -> bool {
   // is being modified by the `where` operator. It would be `MyInterface` in
   // `MyInterface where .Member = i32`.
   auto [self_node, self_id] = context.node_stack().PopExprWithNodeId();
-  auto self_type_id = ExprAsType(context, self_node, self_id).type_id;
+  auto self_with_constraints_type_id =
+      ExprAsType(context, self_node, self_id).type_id;
   // Only facet types may have `where` restrictions.
-  if (self_type_id != SemIR::ErrorInst::TypeId &&
-      !context.types().IsFacetType(self_type_id)) {
+  if (self_with_constraints_type_id != SemIR::ErrorInst::TypeId &&
+      !context.types().IsFacetType(self_with_constraints_type_id)) {
     CARBON_DIAGNOSTIC(WhereOnNonFacetType, Error,
                       "left argument of `where` operator must be a facet type");
     context.emitter().Emit(self_node, WhereOnNonFacetType);
-    self_type_id = SemIR::ErrorInst::TypeId;
+    self_with_constraints_type_id = SemIR::ErrorInst::TypeId;
+  }
+
+  // Strip off any constraints provided by a `WhereExpr` from the `Self` facet
+  // type. For a facet type like `I & J where .X = .Y`, this will reduce it down
+  // to just `I & J`.
```

Comment:
Does this allow:
```carbon
interface I { let X:! type; Y:! type; }
class C(T:! I where .X = ()) {}
fn F(U:! (I where .X = ()) where .Y = C(.Self));
```
In particular, if we're now stripping the `where .X = ()` from `.Self`, does that mean that we no longer allow that to be used when type-checking uses of `.Self` on the right of the `where`?

Reply:
In fact, because `F`, `G`, and `H` are functions, I think we accept them all regardless due to the "implied constraints" logic -- `U.X == ()` would become an implied constraint of the function if we can't prove it when checking the `where` expression. So perhaps it'd be better to use an example in a non-implied-constraints context:
```carbon
interface J {
  let U:! (I where .X = ()) where .Y = C(.Self);
  let V:! I where .X = () and .Y = C(.Self);
  let W:! I where .Y = C(.Self) and .X = ();
}
```